### PR TITLE
refactor: move CRD registration to separate cell

### LINKED ISSUE #24068: Modularization: Operator CRDs registration
**What**

At startup, the operator registers all the CRDs with the k8s api server. It does that calling `client.RegisterCRDs`.

To allow cleaner separation of concerns, it is possible to develop a cell the invokes the CRDs registration.
This should be done just once after the operator has been elected leader. Since the `Invoke` calls are executed in order by the hive, this means that the CRDs registration invoke should be the first one to be added in `WithLeaderLifecycle`.

A further improvement may be to refactor what is currently done in the agent to wait for the registration of the CRDs.

**Success Criteria**

- The CRDs registration is moved into an independent `cell.Invoke`
- The CRDs registration should be the first operation to be invoked after the operator is elected leader.
- If the operation is not carried out synchronously, we should verify that the log is not flooded with errors due to temporary missing CRDs.

Scope: `operator/cmd/root.go`

Related: #23425

### PR DESCRIPTION
**Pull request by knight42**
Fixes: #24068 

/cc @pippolo84

### CODE REVIEW

**Comment by christarazi** [comment]
Travis has a failure, but it's unclear to me if it's related to this PR:

```
panic: *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Service].Events() called from /home/travis/gopath/src/github.com/cilium/cilium/operator/pkg/lbipam/lbipam.go:192 has a broken event handler that did not call Done() before event was garbage collected

                                                                                                                                                                                                                                                                                        

goroutine 5 [running]:                                                                                                                                                                                                                                                                  

github.com/cilium/cilium/pkg/k8s/resource.(*resource[...]).Events.func1.1()                                                                                                                                                                                                             

	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/resource/resource.go:343 +0x68                                                                                                                                                                                                 

FAIL	github.com/cilium/cilium/operator/pkg/lbipam	3.698s                        
```

cc @dylandreimerink

File: operator/cmd/root.go
Comment (pippolo84) [inline operator/cmd/root.go]: We should avoid carrying out blocking operations (like `client.RegisterCRDs`) in the `cell.Invoke` itself. 

What we should do instead is:
- declare a `cell.Module` variable in a new `cell.go` file inside `pkg/k8s/api/cilium.io/client` (this will improve the code files structure).
- In `cell.Invoke`, register a callback to the lifecycle start hook, and run the CRD registration there.

You can look at the identitygc cell [here](https://github.com/knight42/cilium/blob/refactor/modularize-crd-registration/operator/identitygc/cell.go#L38-L46) and [here](https://github.com/knight42/cilium/blob/refactor/modularize-crd-registration/operator/identitygc/gc.go#L107-L126) as an example (even if in this case, the stop hook will do nothing).

File: operator/cmd/root.go
Comment (pippolo84) [inline operator/cmd/root.go]: We can improve the readability using early return:

```
if !clientset.IsEnabled() || operatorCfg.SkipCRDCreation {
  log.Info("Skipping ...")
  return
}

if err := client.RegisterCRDs(...); ...
```

File: operator/cmd/root.go
Comment (pippolo84) [inline operator/cmd/root.go]: When using `cell.Invoke` it is sufficient to return the error and let the caller manage it (the hive will report it and stop).

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for your PR!
I left some comments inline with suggestions on how to restructure the new cell.

**Comment by knight42** [comment]
> Thanks for your PR! I left some comments inline with suggestions on how to restructure the new cell.

Thanks for your detailed guidance, I will update this PR in a few days.

File: operator/cmd/root.go
Comment (knight42) [reply operator/cmd/root.go]: I have moved the cell definition to `pkg/k8s/apis/cilium.io/client/cell.go`, PTAL

**Comment by knight42** [comment]
> Travis has a failure, but it's unclear to me if it's related to this PR:
> 
> ```
> panic: *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1.Service].Events() called from /home/travis/gopath/src/github.com/cilium/cilium/operator/pkg/lbipam/lbipam.go:192 has a broken event handler that did not call Done() before event was garbage collected
> 
>                                                                                                                                                                                                                                                                                         
> 
> goroutine 5 [running]:                                                                                                                                                                                                                                                                  
> 
> github.com/cilium/cilium/pkg/k8s/resource.(*resource[...]).Events.func1.1()                                                                                                                                                                                                             
> 
> 	/home/travis/gopath/src/github.com/cilium/cilium/pkg/k8s/resource/resource.go:343 +0x68                                                                                                                                                                                                 
> 
> FAIL	github.com/cilium/cilium/operator/pkg/lbipam	3.698s                        
> ```
> 
> cc @dylandreimerink

The test passed this time, seems to be a flaky one 🤔

File: pkg/k8s/apis/cilium.io/client/cell.go
Comment (pippolo84) [inline pkg/k8s/apis/cilium.io/client/cell.go]: Since the package name `client` is quite generic, probably it is better to be more specific with the name of the Cell. Something like `RegisterCRDsCell` should do.

File: operator/cmd/root.go
Comment (pippolo84) [inline operator/cmd/root.go]: This cell needs to be the first one to be executed once the operator instance is elected leader (ref: https://github.com/cilium/cilium/issues/24068). It'd be nice to add a comment to highlight that.

File: operator/cmd/root.go
Comment (pippolo84) [inline operator/cmd/root.go]: I think we can get rid of the "SharedConfig" trick. Usually we want that to break the direct dependency of a cell config to global options.

But in this case, the `SkipCRDCreation` option is only used by the new cell, so it is not "global" anymore.
Thus we can remove that option from the `operatorOption.OperatorConfig` struct (alongside any reference to it in `operator/cmd/fags.go`, `operator/option/config.go` and `pkg/option/config.go`) and let the cell declare and initialize it through `cell.Config`.

Something similar can be seen [here](https://github.com/cilium/cilium/blob/master/clustermesh-apiserver/health.go#L19-L33) for the `ClusterMeshHealthPort` option (even though here the declaration of the option name and its default value have been kept in packages `option` and `defaults` respectively).

Please note that this will break the controlplane test setup in `test/controlplane/testcase.go`, where we force-skip the CRDs registration setting `operatorOption.Config.SkipCRDCreation = true` in `*ControlPlaneTest.SetupEnvironment()`.

To fix this you can add something like this:

```
   vp := cpt.operatorHandle.hive.Viper()
   vp.Set(client.SkipCRDCreation, true)
```

in `*ControlPlaneTest.StartOperator()`, just after the operator hive instantiation.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Thanks for the updated PR!

The cell is ok, my main concern is about the management of the `skipCRDCreation` option. We can improve it moving to the cell itself.

File: operator/cmd/root.go
Comment (knight42) [reply operator/cmd/root.go]: Really appreciate the detailed guidance 👍 I have updated the PR, please take a second look

File: operator/cmd/flags.go
Comment (knight42) [inline operator/cmd/flags.go:287]: @pippolo84 just want to double check, should I invoke `option.BindEnv(Vp, client.SkipCRDCreation)` here to bind the env var?

Comment (pippolo84) [reply operator/cmd/flags.go:287]: No, it will be handled by `cell.Config`.

File: test/controlplane/suite/testcase.go
Comment (pippolo84) [inline test/controlplane/suite/testcase.go]: Here the `cpt.operatorHandle` is still `nil`, since the operator has not yet been initialized. This line should be moved in `StartOperator`, just after the initialization of the operator handle.

**Review by pippolo84 — CHANGES_REQUESTED** [review]
Changes look good, only things remaining are:

- fix the operator options setting in controlplane test
- update documentation in cmdref

I left some notes inline.

File: pkg/k8s/apis/cilium.io/client/cell.go
Comment (pippolo84) [inline pkg/k8s/apis/cilium.io/client/cell.go:19]: Now that the option has been moved, the cmdref documentation needs to be rebuilt. You can do that simply running:

`make -C Documentation/ update-cmdref`

**Comment by knight42** [comment]
@pippolo84 Hi I believe all the comments have been addressed, please take another look.

**Review by pippolo84 — APPROVED** [review]
LGTM! Thanks again for the PR and all the further iterations! :100:

**Comment by knight42** [comment]
> LGTM! Thanks again for the PR and all the further iterations! 💯

I am really grateful to your patience and kindness 👍

**Comment by knight42** [comment]
@pippolo84 The failed tests are due to cancelation, looks like they were stuck somewhere for about 1 hour. How could I rerun these tests?

**Comment by pippolo84** [comment]
/test-1.25-4.19

**Comment by pippolo84** [comment]
> @pippolo84 The failed tests are due to cancelation, looks like they were stuck somewhere for about 1 hour. How could I rerun these tests?

I've restarted the canceled tests, let's see how it goes.

**Comment by knight42** [comment]
@pippolo84 There is one remaining failed test `k8s-1.16-kernel-4.19 (test-1.16-4.19)`, the error also seems unrelated:
```
/home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.19/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:515
migrate-svc restart count values do not match
Expected
    <int>: 0
to be identical to
    <int>: 5
/home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.19/src/github.com/cilium/cilium/test/k8s/updates.go:346
```

**Comment by pippolo84** [comment]
> @pippolo84 Got it, so.. could we simply rerun this test? If all tests become green, I think we could get this PR merged +1

We usually avoid rerunning in this case, it would just hide the flake.
For the PR to be marked as ready to merge, we should wait for the other reviewers feedback.

File: Documentation/cmdref/cilium-operator-alibabacloud_hive.md
Comment (christarazi) [inline Documentation/cmdref/cilium-operator-alibabacloud_hive.md:29]: Please squash this commit with the previous commit which caused these changes

File: test/controlplane/suite/testcase.go
Comment (christarazi) [inline test/controlplane/suite/testcase.go:171]: Please squash this commit with the original commit that introduced this change

**Review by christarazi — CHANGES_REQUESTED** [review]

**Comment by knight42** [comment]
@christarazi I have squashed the commits, PTAL

**Comment by knight42** [comment]
gently ping @christarazi

**Review by christarazi — APPROVED** [review]

**Comment by youngnick** [comment]
@knight42 the outstanding CI failure is a known flake that should now be fixed, #23272. If you rebase, it should clear up, and then this should be able to be merged.

**Comment by knight42** [comment]
> @knight42 the outstanding CI failure is a known flake that should now be fixed, #23272. If you rebase, it should clear up, and then this should be able to be merged.

Hi I have rebased and there is no CI error this time 🎉

**Review by rolinh — APPROVED** [review]

**Comment by christarazi** [comment]
@pippolo84 We should track the CI failures before we mark ready to merge.

**Comment by pippolo84** [comment]
/test-1.25-4.19

**Comment by christarazi** [comment]
/test-1.25-4.19

**Comment by christarazi** [comment]
/test-1.24-5.4

**Pull request merged**
